src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs

```diff
@@ -492,15 +499,28 @@ private static int GetDistinctNumberOfSources(TestRunCriteria testRunCriteria)
             // Check the user parallel setting.
             int userParallelSetting = RunSettingsUtilities.GetMaxCpuCount(runSettings);
             parallelLevelToUse = userParallelSetting == 0
-                // TODO: use environment helper so we can control this from tests.
-                ? Environment.ProcessorCount
+                ? _environment.ProcessorCount
                 : userParallelSetting;
-            var enableParallel = parallelLevelToUse > 1;
 
             EqtTrace.Verbose(
                 "TestEngine: Initializing Parallel Execution as MaxCpuCount is set to: {0}",
                 parallelLevelToUse);
 
+            // TODO: EXPERIMENTAL FEATURE - will need to be removed or strengthen/tested.
+            // A negative value is used to indicate that the value should be used as a percentage of the number of cores.
+            if (parallelLevelToUse < 0)
+            {
+                Debug.Assert(parallelLevelToUse > -100, "Parallel level should be above -100 (i.e. 100% of cores)");
+                parallelLevelToUse = (int)Math.Round((double)-parallelLevelToUse * _environment.ProcessorCount / 100.0);
```

Comment:
```suggestion
                // Use the given percent of available cores, while rounding up. Always use at least 1 core.
                Debug.Assert(parallelLevelToUse > -100, "Parallel level should be above -100 (i.e. 100% of cores)");
                parallelLevelToUse = (int)Math.Max(1, Math.Round((double)-parallelLevelToUse * _environment.ProcessorCount / 100.0, MidpointRounding.AwayFromZero));
```

Reply:
The if below should be removed it this change is used. 

You could also remove that Debug.Assert, because it will only annoy us, and use Math.Min on the result with the count of logical cores, so we never say to use more than what we have.